nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.io.FileIO;
+
+/**
+ * Nessie implementation of Iceberg TableOperations.
+ */
+public class NessieTableOperations extends BaseMetastoreTableOperations {
+
+  private final NessieClient client;
+  private final ContentsKey key;
+  private UpdateableReference reference;
+  private IcebergTable table;
+  private FileIO fileIO;
+
+  /**
+   * Create a nessie table operations given a table identifier.
+   */
+  public NessieTableOperations(
+      ContentsKey key,
+      UpdateableReference reference,
+      NessieClient client,
+      FileIO fileIO) {
+    this.key = key;
+    this.reference = reference;
+    this.client = client;
+    this.fileIO = fileIO;
+  }
+
+  @Override
+  protected void doRefresh() {
+    try {
+      reference.refresh();
+    } catch (NessieNotFoundException e) {
+      throw new RuntimeException("Failed to refresh as ref is no longer valid.", e);
+    }
+    String metadataLocation = null;
+    try {
+      Contents contents = client.getContentsApi().getContents(key, reference.getHash());
+      this.table = contents.unwrap(IcebergTable.class)
+          .orElseThrow(() ->
+              new IllegalStateException("Cannot refresh iceberg table: " +
+                  String.format("Nessie points to a non-Iceberg object for path: %s.", key)));
+      metadataLocation = table.getMetadataLocation();
+    } catch (NessieNotFoundException ex) {
+      if (currentMetadataLocation() != null) {
+        throw new NoSuchTableException(ex, "No such table %s", key);
+      }
+    }
+    refreshFromMetadataLocation(metadataLocation, 2);
+  }
+
+  @Override
+  protected void doCommit(TableMetadata base, TableMetadata metadata) {
+    reference.checkMutable();
+
+    String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);
+
+    boolean threw = true;
+    try {
+      IcebergTable newTable = ImmutableIcebergTable.builder().metadataLocation(newMetadataLocation).build();
+      client.getContentsApi().setContents(key,
+                                          reference.getAsBranch().getName(),
+                                          reference.getHash(),
+                                          String.format("iceberg commit%s", applicationId()),
+                                          newTable);
+      threw = false;
+    } catch (NessieConflictException ex) {
+      String fixMsg = reference.isBranch() ?
+          String.format("Update the reference %s and try again", reference.getName()) :
+          String.format("Can't commit to the tag %s", reference.getName());
+      throw new CommitFailedException(ex, "Commit failed: Reference hash is out of date. %s", fixMsg);
+    } catch (NessieNotFoundException ex) {
+      throw new RuntimeException(String.format("Commit failed: Reference %s does not exist", reference.getName()), ex);
```

Comment:
Like above, do we know this is the ref because the current ref already loaded the table?

Reply:
I think my comment above covers this as well. The only sensible way this could ever throw is if someone deleted your branch between when you loaded the table and when you tried to commit.

I have made that a bit more explicit in the comment